diff --git a/CMakeLists.txt b/CMakeLists.txt
index 941865f..0103498 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -89,7 +89,7 @@ set(TON_ARCH "native" CACHE STRING "Architecture, will be passed to -march=")
 
 if (TON_USE_ABSEIL)
   message("Add abseil-cpp")
-  add_subdirectory(third-party/abseil-cpp EXCLUDE_FROM_ALL)
+  find_package(absl REQUIRED)
   set(ABSL_FOUND 1)
 endif()
 
@@ -108,7 +108,7 @@ set(CRC32C_BUILD_BENCHMARKS OFF CACHE BOOL "Build CRC32C's benchmarks")
 set(CRC32C_USE_GLOG OFF CACHE BOOL "Build CRC32C's tests with Google Logging")
 set(CRC32C_INSTALL OFF CACHE BOOL "Install CRC32C's header and library")
 message("Add crc32c")
-add_subdirectory(third-party/crc32c EXCLUDE_FROM_ALL)
+find_package(Crc32c REQUIRED)
 set(CRC32C_FOUND 1)
 
 if (TON_USE_ROCKSDB)
@@ -120,7 +120,7 @@ if (TON_USE_ROCKSDB)
   set(WITH_TOOLS OFF CACHE BOOL "build with tools")
   set(FAIL_ON_WARNINGS OFF CACHE BOOL "fail on warnings")
   message("Add rocksdb")
-  add_subdirectory(third-party/rocksdb EXCLUDE_FROM_ALL)
+  find_package(RocksDB REQUIRED)
 endif()
 
 option(USE_COROUTINES "experimental support of coroutines" OFF)
