diff --git a/.cargo/config.offline b/.cargo/config.offline
index ca6c2b3c3..9e384e09b 100644
--- a/.cargo/config.offline
+++ b/.cargo/config.offline
@@ -9,7 +9,7 @@ git = "https://github.com/zcash/librustzcash.git"
 rev = "d14e7a707ce01cefcbc82651dad48f002185dded"
 replace-with = "vendored-sources"
 
-[source."https://github.com/nuttycom/hdwallet.git"]
+[source."https://github.com/nuttycom/hdwallet"]
 git = "https://github.com/nuttycom/hdwallet.git"
 rev = "9b4c1bdbe0517e3a7a8f285d6048a37d472ba3bc"
 replace-with = "vendored-sources"
