diff --git a/src/core_write.cpp b/src/core_write.cpp
index 097da5722..686e34ad8 100644
--- a/src/core_write.cpp
+++ b/src/core_write.cpp
@@ -293,7 +293,7 @@ void TxToJSON(const CTransaction& tx,
         ScriptPublicKeyToJSON(txout.scriptPubKey, true, utxoAfterGenesis, entry);
         entry.writeEndObject(false);
 
-        entry.writeEndObject(txout != tx.vout.back());
+        entry.writeEndObject(i != tx.vout.size() - 1);
     }
 
     entry.writeEndArray();
