diff --git a/src/chiabls/src/threshold.cpp b/src/chiabls/src/threshold.cpp
index bcd544bc7..9243c222b 100644
--- a/src/chiabls/src/threshold.cpp
+++ b/src/chiabls/src/threshold.cpp
@@ -2,6 +2,8 @@
 // Distributed under the MIT/X11 software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+#include <memory>
+
 #include "threshold.hpp"
 
 #include "schemes.hpp"
diff --git a/src/httpserver.cpp b/src/httpserver.cpp
index 54ff92d4b..ef5e36dba 100644
--- a/src/httpserver.cpp
+++ b/src/httpserver.cpp
@@ -21,6 +21,7 @@
 #include <sys/stat.h>
 #include <signal.h>
 #include <future>
+#include <deque>
 
 #include <event2/thread.h>
 #include <event2/buffer.h>
